source/application/AMPIC887DriverServer/AMPIC887Controller.cpp

```diff
@@ -61,10 +114,954 @@ QString AMPIC887Controller::status() const
+	}
+
+	return true;
+}
+
+bool AMPIC887Controller::haltSmoothly(const AMPIC887AxisCollection& axes)
+{
+	AMGCS2HaltSmoothlyCommand haltCommand(axes);
+	runCommand(&haltCommand);
+
+	if(haltCommand.runningState() != AMGCS2Command::Succeeded) {
+		setError(haltCommand.lastError());
+		return false;
+	}
+
+	return true;
+}
+
+void AMPIC887Controller::move(const AMPIC887AxisMap<double>& axisPositions)
+{
+	AMPIC887AxisCollection axesToMove = axisPositions.axes();
+	foreach(AMGCS2::Axis currentAxis, axesToMove) {
+		if(!controllerState_->hexapodState()->referencedState(currentAxis)) {
+			setError(QString("Cannot move axis %1: Axis not referenced")
+					 .arg(AMGCS2Support::axisToCharacter(currentAxis)));
+			return;
+		}
+	}
+
+	AMGCS2AsyncMoveCommand* asyncMoveCommand = new AMGCS2AsyncMoveCommand(axisPositions);
+
+	connect(asyncMoveCommand, SIGNAL(started(AMGCS2AsyncCommand*)),
+			this, SLOT(onAsyncMoveStarted(AMGCS2AsyncCommand*)));
+	connect(asyncMoveCommand, SIGNAL(failed(AMGCS2AsyncCommand*)),
+			this, SLOT(onAsyncMoveFailed(AMGCS2AsyncCommand*)));
+	connect(asyncMoveCommand, SIGNAL(succeeded(AMGCS2AsyncCommand*)),
+			this, SLOT(onAsyncMoveSucceeded(AMGCS2AsyncCommand*)));
+
+	runCommand(asyncMoveCommand);
+}
+
+void AMPIC887Controller::moveRelative(const AMPIC887AxisMap<double>& relativePositions)
+{
+	AMPIC887AxisCollection axesToMove = relativePositions.axes();
+	foreach(AMGCS2::Axis currentAxis, axesToMove) {
+		if(!controllerState_->hexapodState()->referencedState(currentAxis)) {
+			setError(QString("Cannot move axis %1: Axis not referenced")
+					 .arg(AMGCS2Support::axisToCharacter(currentAxis)));
+			return;
+		}
+	}
+
+	AMGCS2AsyncMoveRelativeCommand* asyncMoveRelativeCommand =
+			new AMGCS2AsyncMoveRelativeCommand(relativePositions);
+
+	connect(asyncMoveRelativeCommand, SIGNAL(started(AMGCS2AsyncCommand*)),
+			this, SLOT(onAsyncMoveRelativeStarted(AMGCS2AsyncCommand*)));
+	connect(asyncMoveRelativeCommand, SIGNAL(failed(AMGCS2AsyncCommand*)),
+			this, SLOT(onAsyncMoveRelativeFailed(AMGCS2AsyncCommand*)));
+	connect(asyncMoveRelativeCommand, SIGNAL(succeeded(AMGCS2AsyncCommand*)),
+			this, SLOT(onAsyncMoveRelativeSucceeded(AMGCS2AsyncCommand*)));
+
+	runCommand(asyncMoveRelativeCommand);
+}
+
+void AMPIC887Controller::referenceMove(const AMPIC887AxisCollection& axes)
+{
+	AMGCS2AsyncReferenceMoveCommand* asyncReferenceMoveCommand =
+			new AMGCS2AsyncReferenceMoveCommand(axes);
+
+	connect(asyncReferenceMoveCommand, SIGNAL(started(AMGCS2AsyncCommand*)),
+			this, SLOT(onAsyncReferenceMoveStarted(AMGCS2AsyncCommand*)));
+	connect(asyncReferenceMoveCommand, SIGNAL(failed(AMGCS2AsyncCommand*)),
+			this, SLOT(onAsyncReferenceMoveFailed(AMGCS2AsyncCommand*)));
+	connect(asyncReferenceMoveCommand, SIGNAL(succeeded(AMGCS2AsyncCommand*)),
+			this, SLOT(onAsyncReferenceMoveSucceeded(AMGCS2AsyncCommand*)));
+
+	runCommand(asyncReferenceMoveCommand);
+}
+
+void AMPIC887Controller::onAsyncMoveStarted(AMGCS2AsyncCommand *command)
```

Comment:
Is this command used for when composite moves (say x and w or x, y, z, u, v) are done?